yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+      logInfo(s"Shutting down: pushing out ${actionQueue.size} events")
+      innerEnqueue(FlushTimelineEvents())
+
+      // stop operation
+      postThreadActive synchronized {
+        innerEnqueue(StopQueue())
+        // now await that marker flag
+        postThreadActive.wait(maxTimeToWaitOnShutdown)
+      }
+
+      if (!actionQueue.isEmpty) {
+        // likely cause is ATS not responding.
+        // interrupt the thread, albeit at the risk of app stop events
+        // being lost. There's not much else that can be done if
+        // ATS is being unresponsive
+        logWarning(s"Did not finish flushing actionQueue before " +
+          s"stopping ATSService, eventQueueBacklog= ${actionQueue.size}")
+        eventHandlingThread.foreach(_.interrupt())
+      }
+      stopTimelineClient
+      logInfo(s"Stopped: $this")
+    }
+  }
+
+  /**
+   * Get the current entity, creating one on demand if needed, setting up its entity Id
+   * from the current appID and attempt ID values, filters from the current name and user
+   * @return the current entity
+   */
+  private def getOrCreateCurrentEntity() = {
+    curEntity.getOrElse {
+      currentEventCount = 0
+      val entity = createTimelineEntity(
+        _applicationId,
+        _attemptId,
+        sparkApplicationId,
+        sparkApplicationAttemptId,
+        appStartEventProcessed.get(),
+        appName,
+        userName)
+      curEntity = Some(entity)
+      logDebug(s"Demand creation of new entity ${describeEntity(entity)}")
+      entity
+    }
+  }
+
+  /**
+   * Perform any preflight checks.
+   * @param entity timeline entity to review.
+   */
+  private[yarn] def preflightCheck(entity: TimelineEntity): Unit = {
+    require(entity.getStartTime != null,
+      s"No start time in ${describeEntity(entity)}")
+  }
+
+  /**
+   * If there is any available entity to be sent, push to timeline server
+   */
+  private def flushEntity(): Unit = {
+    if (entityList.nonEmpty) {
+      val count = flushCount.incrementAndGet()
+      logDebug(s"flushEntity #$count: list size ${entityList.size}")
+      var client = getTimelineClient
+      val lastUpdated = now()
+      // attempt to post each entity, removing from the list all of those
+      // which were successfully posted. The remainder are retained for
+      // followon attempts
+      entityList = entityList.filter { entity => {
+          if (entity == null) {
+            false
+          } else {
+            entity.addOtherInfo(FIELD_LAST_UPDATED, lastUpdated)
+            if (domainId != null) {
+              entity.setDomainId(domainId)
+            }
+            val entityDescription = describeEntity(entity)
+            logDebug(s"About to POST $entityDescription")
+            try { 
+              val response: TimelinePutResponse = client.putEntities(entity)
+              val errors = response.getErrors
+              if (!response.getErrors.isEmpty) {
+                val err = errors.get(0)
```

Comment:
IIRC, you can only get one error here because you're just sending one entity in the `putEntities` call, right? If that's the case, the code is correct but a little odd (since it records `errors.size()` instead of `1` to indicate you know there's a single error), so maybe a comment?